richtextfx/src/main/java/org/fxmisc/richtext/model/Paragraph.java

```diff
@@ -132,7 +132,8 @@ public String substring(int from) {
             int segIdx = pos.getMajor();
             List<SEG> segs = new ArrayList<>(segIdx + 1);
             segs.addAll(segments.subList(0, segIdx));
-            segs.add(segmentOps.subSequence(segments.get(segIdx), 0, pos.getMinor()));
+            segmentOps.subSequence(segments.get(segIdx), 0, pos.getMinor())
+                            .ifPresent(seg -> segs.add(seg));
```

Comment:
> If we wrap the returned value of subsequence in an Optional, it adds boilerplate to the code (and is honestly so confusing that I'm not even sure if I've implemented it correctly, nor would I want to ever have to read through this code).

Indeed - I also tried that last week and gave up at the `Paragraph.restyle()` methods - I **was** pretty sure that I would not be able to implement that correctly :joy: 

I'll look into your PR and let you know my progress...